### ISSUE TITLE
Problem with recording both speakers and mic

### ISSUE THREAD

**Issue by mao73a** [AntonyCorbett/OnlyR#93]
**Describe the bug**
ONLYR records either mic or a sound output. I am unable to record both mic and speakers output during calls.

**To Reproduce**
1. Go to settings, check "Use loop"
2. Click on Record
3. Only windows output sounds are recorded (only other people voices are recorded during a call but not mine)
4. Go to settings, un-check "Use loop"
5. Click on Record
6. Only my mic is recorded - not other people voices

**Expected behavior**
I would expect the app to record both my mic and other peeople voices during a call.

**Desktop (please complete the following information):**
 - Windows 11 Pro 23H2
 - Sound device: Realtek High Definition Audio

**Comment by rubenmarchan** [comment]
I second this request @AntonyCorbett @WailGree  do you think we can prioritize this update?

**Comment by WailGree** [comment]
@rubenmarchan I actually started working on this already.
The only thing left is to merge the audios of the microphone and system into one.
@AntonyCorbett How do you think this should be presented on the UI?
I was thinking of maybe a drop-down instead of the "capture loopback" checkbox.
"Audio source:"
With the following options:
"Input",
"Loopback",
"Input & Loopback" or "Both"

**Comment by rubenmarchan** [comment]
@AntonyCorbett @WailGree I have an example to illustrate this. There’s another project called StepVoice Recorder, but it was abandoned by its previous developer. The developer had implemented a checkbox menu that enabled users to select multiple speakers and microphones, and the user could then make their selections.

**Comment by AntonyCorbett** [comment]
> ...enabled users to select multiple speakers and microphones...

Is this technically within the scope of your work, e.g. to select multiple microphones?

**Comment by rubenmarchan** [comment]
There is a use case that StepVoice Recorder was used to record calls. The user will select the microphone and speaker of the headset and its a way to record calls.

**Comment by rubenmarchan** [comment]
@AntonyCorbett @WailGree any updates on this?

**Comment by haili-WAN** [comment]
I also look forward to this feature. We need to record the sound of the computer itself and the sound of the headphone device talking in the Windows system.

**Comment by ryrylightfoot** [comment]
I just wanted to bump this issue. We can't use Only-R to record our meetings. We need to record the mics, Zoom, and the media played on the PC. Before the pandemic I don't remember an issue recording the media and the mics, did Windows 11 mess something up with the sound? We're instructed not to use Zoom to record, so that's not an option. I hope this gets added to OnlyR. Thank you for all your work!

**Comment by AntonyCorbett** [comment]
> I just wanted to bump this issue. We can't use Only-R to record our meetings. We need to record the mics, Zoom, and the media played on the PC. Before the pandemic I don't remember an issue recording the media and the mics, did Windows 11 mess something up with the sound? We're instructed not to use Zoom to record, so that's not an option. I hope this gets added to OnlyR. Thank you for all your work!

I think this may be a different problem from that described in this issue. I guess you are discussing the use of OnlyR at congregation meetings rather than for conference calls. In your case, you just need to select the correct audio input - the one that provides audio out from your mixer.

**Comment by ryrylightfoot** [comment]
> > I just wanted to bump this issue. We can't use Only-R to record our meetings. We need to record the mics, Zoom, and the media played on the PC. Before the pandemic I don't remember an issue recording the media and the mics, did Windows 11 mess something up with the sound? We're instructed not to use Zoom to record, so that's not an option. I hope this gets added to OnlyR. Thank you for all your work!
> 
> I think this may be a different problem from that described in this issue. I guess you are discussing the use of OnlyR at congregation meetings rather than for conference calls. In your case, you just need to select the correct audio input - the one that provides audio out from your mixer.

Thank you for replying. The issue is definitely the same. When I have the mic selected, Only-R records all the audio from the mixer, but nothing from Windows; no Zoom, no media. When I check the loopback capture, it records Zoom and the media, but nothing from the mixer.

**Comment by AntonyCorbett** [comment]
> Thank you for replying. The issue is definitely the same. When I have the mic selected, Only-R records all the audio from the mixer, but nothing from Windows; no Zoom, no media. When I check the loopback capture, it records Zoom and the media, but nothing from the mixer.

Thanks for your reply. This is a configuration issue. Audio recording works fine in a KH setting without the use of loopback. Loopback simply records all the audio being sent out of your PC soundcard. You don't need this for a KH meeting.

You say that when you select "mic" input in OnlyR it records all the audio from the mixer, but then you indicate that it doesn't record audio from the PC. This must mean that the PC audio is not being sent to the mixer, or else you are not including it in the output bus sent back to the PC. There are several types of configuration depending on the equipment in use (see S-261).

This issue (#39) is about using OnlyR to record conference calls. Would you mind creating another issue to follow up on your problem if needed? Thanks, Antony

**Comment by ryrylightfoot** [comment]
We use the recommended Behringer U-Control UCA222 USB Audio Interface and a TOA A-812. Everything is set up as instructed. It was recently remodeled. The PC sound is not sent back to the PC.

**Comment by AntonyCorbett** [comment]
> We use the recommended Behringer U-Control UCA222 USB Audio Interface and a TOA A-812. Everything is set up as instructed. It was recently remodeled. The PC sound is not sent back to the PC.

Is the REC (recording) output of your TOA mixer connected to the input of your UCA222? This makes the output of your mixer available for recording at the PC via the UCA222 "mic" device.

**Comment by ryrylightfoot** [comment]
> Is the REC (recording) output of your TOA mixer connected to the input of your UCA222? This makes the output of your mixer available for recording at the PC via the UCA222 "mic" device. If it is connected (and the fact that you can record your mics successfully would indicate it is), then the only other thing I can think is that the audio output on your PC is being sent to somewhere other than the UCA222, e.g. the speakers in your monitors.

Yes, the cords are all as they should be. We have no issues during the meetings. Only with recording. The rec out goes to the UCA222 in and the UCA222 out goes to the input on the mixer.

**Comment by AntonyCorbett** [comment]
@ryrylightfoot if all audio signals from your mixer arrive at the PC then you should be able to record them on the single audio input device. There is nothing unusual about this configuration and OnlyR is routinely used to record in this environment. The issue you are experiencing is not the same as that reported here. If you want to continue a discussion please create a separate issue and ask for help from people who have a similar setup. Thanks, Antony

**Comment by rubenmarchan** [comment]
@WailGree @AntonyCorbett any updates on this?

**Comment by AntonyCorbett** [comment]
@rubenmarchan Yes, I have made some progress, but not enough to build something to test yet

**Comment by AntonyCorbett** [comment]
@rubenmarchan @mao73a @WailGree @haili-WAN Please see the latest pre-release. This is an experimental release based on my dual-input-2 branch. Please provide feedback.

**Comment by AntonyCorbett** [comment]
@rubenmarchan thanks for your feedback. The implementation allows for recording from a specified device and the loopback capture - per the original requirement

**Comment by rubenmarchan** [comment]
[log-20250224.txt](https://github.com/user-attachments/files/20765747/log-20250224.txt)
[log-20250616.txt](https://github.com/user-attachments/files/20765746/log-20250616.txt)

**Comment by AntonyCorbett** [comment]
@rubenmarchan Thanks for providing the log file. The two audio sources needed to have the same sample rate and channel count. I've now fixed this in release 2.2.0.2 - please would you test? I have limited audio devices to test and have not been able to reproduce the issue, but the log file error report is clear as to the cause. Thanks for your help. Antony

**Comment by rubenmarchan** [comment]
Hi @AntonyCorbett  , the release works, but the loopback capture is picking up the audio from the computer’s audio card. For instance, I’m using a headset with both microphone and speaker. I want to listen to the person on the call through the headset and record the audio coming from the headset’s microphone as well as my own voice. Is there a way to achieve this?

Currently I had to switch the audio to come out of the laptop for OnlyR to record the audio. If I switch the audio to come out on the headset the software does not pick up the audio

**Comment by rubenmarchan** [comment]
@AntonyCorbett I am adding a follow up comment. I made the headset the default “speaker” in Windows and I was able to achieve recording both, and listening on the headset. So I think this release version could work.

However adding a way to select the loop back device similar to team would be great.

I am attaching a picture of how Teams shows the Audio options.

![Image](https://github.com/user-attachments/assets/48e719e9-1756-4d00-9b43-b0ee1cb8ad3f)
